src/main/java/net/neoforged/neoforge/network/codec/NeoForgeStreamCodecs.java

```diff
@@ -91,6 +92,29 @@ public void encode(B buf, ResourceKey<? extends Registry<?>> value) {
         };
     }
 
+    /**
+     * Creates a stream codec that uses different implementation depending on the {@link net.neoforged.neoforge.network.connection.ConnectionType}.
+     * Should be used to keep vanilla connection compatibility.
+     */
+    public static <V> StreamCodec<RegistryFriendlyByteBuf, V> connectionAware(
+            StreamCodec<? super RegistryFriendlyByteBuf, V> neoForgeCodec,
+            StreamCodec<? super RegistryFriendlyByteBuf, V> otherCodec) {
+        return new StreamCodec<>() {
+            @Override
+            public V decode(RegistryFriendlyByteBuf buf) {
+                return buf.getConnectionType().isNeoForge() ? neoForgeCodec.decode(buf) : otherCodec.decode(buf);
```

Comment:
This can also use a switch statement:
```suggestion
                return switch (buf.getConnectionType()) {
                    case NEOFORGE -> neoForgeCodec.decode(buf);
                    case OTHER -> otherCodec.decode(buf);
                };
```